### PR TITLE
[RFC] add interface for evaluating messages

### PR DESCRIPTION
**Pull request by agola11**
still a wip

### CODE REVIEW

File: langchain/evaluation/run_evaluators/string_run_evaluator.py [hunk 1]
```diff
@@ -315,43 +342,51 @@ async def aevaluate_run(
...
+            elif evaluator.requires_reference:
+                raise ValueError(
+                    f"Evaluator {evaluator.evaluation_name} requires a reference"
+                    " example from the dataset. Please specify the reference key from"
+                    " amongst the dataset outputs keys."
+                )
+            else:
+                example_mapper = None
+        elif isinstance(evaluator, MessageEvaluator):
+            raise NotImplementedError()
```
Comment (agola11) [inline langchain/evaluation/run_evaluators/string_run_evaluator.py:383]: will fill this in

File: langchain/evaluation/run_evaluators/message_run_evaluator.py [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+
+    @staticmethod
+    def extract_inputs(inputs: Dict) -> List[BaseMessage]:
+        if not inputs.get("messages"):
+            raise ValueError("Run must have messages as inputs.")
+        if "messages" in inputs:
+            if isinstance(inputs["messages"], list) and inputs["messages"]:
+                if isinstance(inputs["messages"][0], BaseMessage):
+                    return messages_from_dict(inputs["messages"])
+                elif isinstance(inputs["messages"][0], list):
```
Comment (agola11) [inline langchain/evaluation/run_evaluators/message_run_evaluator.py:95]: even if we don't merge this is PR in, we should cherry pick this to the LLMStringRunMapper cc @hinthornw @nfcampos

File: langchain/evaluation/run_evaluators/string_run_evaluator.py [hunk 3]
```diff
@@ -28,10 +28,10 @@ def output_keys(self) -> List[str]:
         return ["prediction", "input"]
 
     @abstractmethod
-    def map(self, run: Run) -> Dict[str, Any]:
+    def map(self, run: Run) -> Dict[str, str]:
```
Comment (agola11) [inline langchain/evaluation/run_evaluators/string_run_evaluator.py:31]: these changes should also be cherry picked

File: langchain/evaluation/run_evaluators/message_run_evaluator.py (hunk 2 above)
Comment (hinthornw) [reply langchain/evaluation/run_evaluators/message_run_evaluator.py:95]: Cherry picked

**Comment by hinthornw** [comment]
I'll cherry pick the fixes to the other branch. I think probably better to hold off on message evaluator since we don't really have a great representative OSS evaluator to use it yet

**Pull request closed without merging**